python/lsst/ctrl/bps/bps_reports.py

```diff
@@ -251,26 +251,26 @@ def add(self, run_report, use_global_id=False):
             id_ = run_report.global_wms_id if use_global_id else run_report.wms_id
             self._msg = f"WARNING: Job summary for run '{id_}' not available, report maybe incomplete."
             return
+
+        # Possible payload/pipetask error codes:
+        # * 1: general failure,
+        # * 2: command line error (e.g. unknown command and/or option).
+        pyld_error_codes = {1, 2}
+        pyld_exit_codes = {0} | pyld_error_codes
+
         exit_code_summary = run_report.exit_code_summary
         for label in labels:
             exit_codes = exit_code_summary[label]
-            if exit_codes:
-                # payload errors always return 1 on failure
-                pipe_error_count = sum([code for code in exit_codes if code == 1])
-                infra_codes = [code for code in exit_codes if code != 0 and code != 1]
-                if infra_codes:
-                    infra_error_count = len(infra_codes)
-                    str_infra_codes = [str(code) for code in infra_codes]
-                    infra_error_codes = ", ".join(sorted(set(str_infra_codes)))
-                else:
-                    infra_error_count = 0
-                    infra_error_codes = "None"
-            else:
-                pipe_error_count = 0
-                infra_error_codes = "None"
-                infra_error_count = 0
-            run = [label]
-            run.extend([pipe_error_count, infra_error_count, infra_error_codes])
+
+            pyld_errors = [str(code) for code in exit_codes if code in pyld_error_codes]
+            pyld_error_count = len(pyld_errors)
+            pyld_error_summary = ", ".join(sorted(set(pyld_errors))) if pyld_errors else "None"
+
+            infra_errors = [str(code) for code in exit_codes if code not in pyld_exit_codes]
```

Comment:
Yes, `exit_codes` is a list of exit codes per job.

Regarding the performance concerns, as the tests below show, even for a list consisting of a million of elements the code executes in a fraction of a second and is as fast as an alternative, "single pass" solution using a for loop and an if statement. So I don't think it is likely it will become a bottleneck during the execution of `bps report` command.

However, I realized that the current version unnecessarily converts every exit code that matches the search criteria to its string representation. Similarly what I did in the tests, I'll change the existing code to do this conversion only for the final, unique set of the codes.

```
$ python
Python 3.11.7 | packaged by conda-forge | (main, Dec 23 2023, 14:43:09) [GCC 12.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from timeit import timeit
>>> setup = '''
... from random import randint
... exit_codes = [randint(0, 10) for _ in range(1_000_000)]
... pyld_error_codes = {1, 2}
... pyld_exit_codes = {0} | pyld_error_codes
... '''
>>> stmt_lc='''
... pyld_errors = [code for code in exit_codes if code in pyld_error_codes]
... pyld_error_count = len(pyld_errors)
... pyld_error_summary = ", ".join(sorted(str(code) for code in set(pyld_errors))) if pyld_errors else "None"
... infra_errors = [code for code in exit_codes if code not in pyld_exit_codes]
... infra_error_count = len(infra_errors)
... infra_error_summary =  ", ".join(sorted(str(code) for code in set(infra_errors))) if infra_errors else "None"
... '''
>>> stmt_if='''
... pyld_errors, pyld_errors_count = set(), 0
... infra_errors, infra_errors_count = set(), 0
... for code in exit_codes:
...   if code == 0:
...     pass
...   elif code in pyld_error_codes:
...     pyld_errors.add(code)
...     pyld_errors_count += 1
...   else:
...     infra_errors.add(code)
...     infra_errors_count += 1
... pyld_error_summary = ", ".join(sorted(str(code) for code in pyld_errors)) if pyld_errors else "None"
... infra_error_summary =  ", ".join(sorted(str(code) for code in infra_errors)) if infra_errors else "None"
... '''
>>> timeit(stmt=stmt_lc, setup=setup, number=100) / 100
0.057453537139954276
>>> timeit(stmt=stmt_if, setup=setup, number=100) / 100
0.05786083667000639
```